phdi/harmonization/hl7.py

```diff
@@ -0,0 +1,289 @@
+        if line != "":
+            next_message += f"{line}\r"
+
+    # Since our loop only adds messages to output when it finds
+    # a line that starts with MSH, the last message would never
+    # be added. So we explicitly add it here.
+    if next_message != "":
+        output.append(next_message)
+
+    return output
+
+
+def _default_hl7_value(
+    message: str, segment_id: str, field_num: int, default_value: str
+):
+    """
+    Default a field value in an HL7 message.
+
+    :param message: String representing the HL7 message use to modify
+      a value
+    :param segment_id: Segment type (MSH, PID, etc)
+    :param field_num: Field to replace in the HL7 message
+    :param default_value: If the selected field is blank, set the
+      field's value to this
+    :return: The HL7 message with default value inserted at the
+      specified segment location, if possible. If not, then the
+      original message.
+    """
+
+    parsed_message: hl7.Message = None
+    try:
+        # The hl7 module requires \n characters be replaced with \r
+        parsed_message = hl7.parse(message.replace("\n", "\r"))
+
+        segment: hl7.Segment = None
+        try:
+            segment = parsed_message.segment(segment_id=segment_id)
+        except KeyError:
+            # If the segment is not found, there is nothing to do
+            return message
+
+        if (
+            segment.extract_field(field_num=field_num) is None
+            or segment.extract_field(field_num=field_num) == ""
+        ):
+            segment.assign_field(value=default_value, field_num=field_num)
+
+    # @TODO: Eliminate logging, raise an exception, document the exception
+    # in the docstring, and make this fit into our new structure of allowing
+    # the caller to implement more robust error handling
+    except Exception:
+        logging.exception(
+            "Exception occurred while cleaning message.  "
+            + "Passing through original message."
+        )
+
+        return message
+
+    return str(parsed_message).replace("\r", "\n")
+
+
+def _normalize_hl7_datetime_segment(
+    message: hl7.Message, segment_id: str, field_list: list
+) -> None:
+    """
+    Utility function used to apply datetime normalization
+    to multiple fields in a segment.
+
+    :param message: The HL7 message, represented as a list
+      of indexable component strings (which is how the HL7 library
+      processes and returns messages)
+    :param segment_id: The HL7-specified segment (ex. PID) to
+      normalize the fields of
+    :param field_list: The list of fields contained in the
+      indexed message component, which are themselves indices to
+      data strings
+    :return: None (overwrites the input segment)
+    """
+    try:
+        for segment in message.segments(segment_id):
+            for field_num in field_list:
+                if len(segment) > field_num and segment[field_num][0] != "":
```

Comment:
It's a little confusing to me what `segment[field_num][0]` is. Could we add a variable with a descriptive name to make it clear?

Reply:
@spencer-skylight Can we change `field_num` to something like component, or is field already the right nomenclature? I want to make sure we're using the right terminology to address Kenneth's request here.

@KennethSkylight Basically this is an HL7 terminology minefield lol. An HL7 message is broken down into specific _segments_, which are things like PID, OBX, etc. A segment is an order-specific "list", in essence, of sub-fields that contain health-related information. I _believe_ the fields within a segment are actually just called _fields_ in a technical sense, but I could be wrong and they could be components. We just want to make sure we're using the right word here so that people familiar with HL7 will know what the variables mean (since we're making that assumption of familiarity for our user base).

Reply:
`field` is the correct nomenclature for this...
@KennethSkylight 's interpretation is correct, and to address his question, the pipe (|) character separates fields within a segment (e.g. name, gender, etc).  The carat (^) character separates components within a field (e.g. first name, last name, etc). There are also other characters that can further divide components into subcomponents - the ampersand (&), and separate repeated field values - the tilde (~).  

So I think the nomenclature is good. To make it slightly more clear what that value means, you could do something like this...
```suggestion
                component_num = 0  # The datetime value is always found in the 1st component
                field_value = segment[field_num][component_num]
                if len(segment) > field_num and field_value != "":
```

Reply:
Well, have to do a modified version to make sure `len(segment) > field_num` but otherwise this is good